File: jscomp/test/js_promise_basic_test.ml
let suites :  Mt.pair_suites ref  = ref []
let test_id = ref 0
let eq loc x y = 
  incr test_id ; 
  suites := 
    (loc ^" id " ^ (string_of_int !test_id), (fun _ -> Mt.Eq(x,y))) :: !suites



open Js_promise

let assert_bool b =
  if b then ()
  else
    raise (Invalid_argument "Assertion Failure.")

let fail _ =
  (* assert_bool false *)
  assert false 

let thenTest () =
  let p = resolve 4 in
  p |> then_ (fun x -> resolve @@ assert_bool (x = 4))

let andThenTest () =
  let p = resolve 6 in
  p |> then_ (fun _ -> resolve (12))
    |> then_ (fun y -> resolve @@ assert_bool (y = 12))

let h = resolve ()

let assertIsNotFound (x : Js_promise.error) = 
  match (function [@bs.open]
  | Not_found -> 0) x with 
  | Some _ -> h
  | _ -> assert false 

(** would be nice to have [%bs.open? Stack_overflow]*)
let catchTest () =
  let p = reject Not_found in
  p |> then_ fail
    |> catch (fun error -> 
      assertIsNotFound error
    )

let orResolvedTest () =
  let p = resolve 42 in
  p |> catch (fun _ -> resolve 22)
    |> then_ (fun value -> resolve @@ assert_bool (value = 42))
    |> catch fail

let orRejectedTest () =
  let p = reject Not_found in
  p |> catch (fun _ -> resolve 22)
    |> then_ (fun value -> resolve @@ assert_bool (value = 22))
    |> catch fail

let orElseResolvedTest () =
  let p = resolve 42 in
  p |> catch (fun _ -> resolve 22)
    |> then_ (fun value -> resolve @@ assert_bool (value = 42))
    |> catch fail

let orElseRejectedResolveTest () =
  let p = reject Not_found in
  p |> catch (fun _ -> resolve 22)
    |> then_ (fun value -> resolve @@ assert_bool (value = 22))
    |> catch fail

let orElseRejectedRejectTest () =
  let p = reject Not_found in
  p |> catch (fun _ -> reject Stack_overflow)
    |> then_ fail
    |> catch (fun error ->
        match (function [@bs.open] Stack_overflow -> 0) error with 
        | Some _ -> h 
        | None -> assert false 
        (* resolve @@ assert_bool (Obj.magic error == Stack_overflow) *))

let resolveTest () =
  let p1 = resolve 10 in
  p1 |> then_ (fun x -> resolve @@ assert_bool (x = 10))

let rejectTest () =
  let p = reject Not_found in
  p |> catch
    (fun error ->
       assertIsNotFound error
       (* resolve @@ assert_bool (Obj.magic error == Not_found) *)
    )

let thenCatchChainResolvedTest () =
  let p = resolve 20 in
  p |> then_ (fun value -> resolve @@ assert_bool (value = 20) )
    |> catch fail

let thenCatchChainRejectedTest () =
  let p = reject Not_found in
  p |> then_ fail
    |> catch (fun error -> 
      assertIsNotFound error
      (* resolve @@ assert_bool (Obj.magic error == Not_found) *))


let allResolvedTest () =
  let p1 = resolve 1 in
  let p2 = resolve 2 in
  let p3 = resolve 3 in
  let promises = [| p1; p2; p3 |] in
  all promises
  |> then_
  (fun resolved ->
     assert_bool (resolved.(0) = 1) ;
     assert_bool (resolved.(1) = 2) ;
     assert_bool (resolved.(2) = 3) ;
     h
  )
  

let allRejectTest () =
  let p1 = resolve 1 in
  let p2 = resolve 3 in
  let p3 = reject Not_found in
  let promises = [| p1; p2; p3 |] in
  all promises
    |> then_ fail
    |> catch (fun error -> assert_bool (Obj.magic error == Not_found) ; h)

let raceTest () =
  let p1 = resolve "first" in
  let p2 = resolve "second" in
  let p3 = resolve "third" in
  let promises = [| p1; p2; p3 |] in
  race promises
  |> then_ (fun resolved -> h)
  |> catch fail

let createPromiseRejectTest () =
  make (fun ~resolve ~reject -> reject Not_found [@bs])
  |> catch (fun error -> assert_bool (Obj.magic error == Not_found); h)

let createPromiseFulfillTest () =
  make (fun ~resolve ~reject:_  -> resolve "success" [@bs])
  |> then_ (fun resolved -> assert_bool (Obj.magic resolved = "success"); h)
  |> catch fail

let () =
  ignore @@ thenTest ();
  ignore @@ andThenTest ();
  ignore @@ catchTest ();
  ignore @@ orResolvedTest ();
  ignore @@ orRejectedTest ();
  ignore @@ orElseResolvedTest ();
  ignore @@ orElseRejectedResolveTest ();
  ignore @@ orElseRejectedRejectTest ();
  ignore @@ thenCatchChainResolvedTest ();
  ignore @@ thenCatchChainRejectedTest ();
  ignore @@ allResolvedTest ();
  ignore @@ allRejectTest ();
  ignore @@ raceTest ();
  ignore @@ createPromiseRejectTest ();
  ignore @@ createPromiseFulfillTest ()

(** TODO: async tests?
*)
let () = 
    (Js.Promise.all2 (Js.Promise.resolve 2, Js.Promise.resolve 3))
    |> Js.Promise.then_ (fun (a,b) -> 
    eq __LOC__ (a,b) (2,3); 
    
    Js.Promise.resolve ()
    )
    |> ignore


;; Js.log (List.length !suites)     
     
;; Js.log "hey"
;; Mt.from_pair_suites __FILE__ !suites

let twop = Js.Promise.resolve 2 
let then_ = Js.Promise.then_
let re = Js.Promise.resolve

;; Mt.from_promise_suites __FILE__
  [
    __LOC__, 
    twop
    |> then_ (fun x -> re @@ Mt.Eq(x,2));
    __LOC__, 
    twop
    |> then_ (fun x -> re @@ Mt.Neq(x,3))        
  ]

File: jscomp/test/js_promise_basic_test.js
'use strict';

var Mt = require("./mt.js");
var List = require("../../lib/js/list.js");
var Caml_array = require("../../lib/js/caml_array.js");
var Js_promise = require("../../lib/js/js_promise.js");
var Caml_exceptions = require("../../lib/js/caml_exceptions.js");

var suites = {
  contents: /* [] */0
};

var test_id = {
  contents: 0
};

function eq(loc, x, y) {
  test_id.contents = test_id.contents + 1 | 0;
  suites.contents = {
    hd: [
      loc + (" id " + String(test_id.contents)),
      (function (param) {
          return {
                  TAG: /* Eq */0,
                  _0: x,
                  _1: y
                };
        })
    ],
    tl: suites.contents
  };
  
}

function assert_bool(b) {
  if (b) {
    return ;
  }
  throw {
        RE_EXN_ID: "Invalid_argument",
        _1: "Assertion Failure.",
        Error: new Error()
      };
}

function fail(param) {
  throw {
        RE_EXN_ID: "Assert_failure",
        _1: [
          "js_promise_basic_test.ml",
          19,
          2
        ],
        Error: new Error()
      };
}

function thenTest(param) {
  var p = Promise.resolve(4);
  var arg1 = function (x) {
    return Promise.resolve(assert_bool(x === 4));
  };
  return p.then(arg1);
}

function andThenTest(param) {
  var p = Promise.resolve(6);
  var arg1 = function (param) {
    return Promise.resolve(12);
  };
  var obj = p.then(arg1);
  var arg1$1 = function (y) {
    return Promise.resolve(assert_bool(y === 12));
  };
  return obj.then(arg1$1);
}

var h = Promise.resolve(undefined);

function assertIsNotFound(x) {
  var match = Caml_exceptions.is_extension(x) && x.RE_EXN_ID === "Not_found" ? 0 : undefined;
  if (match !== undefined) {
    return h;
  }
  throw {
        RE_EXN_ID: "Assert_failure",
        _1: [
          "js_promise_basic_test.ml",
          36,
          9
        ],
        Error: new Error()
      };
}

function catchTest(param) {
  var p = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var obj = p.then(fail);
  var arg1 = assertIsNotFound;
  return obj.catch(arg1);
}

function orResolvedTest(param) {
  var p = Promise.resolve(42);
  var arg1 = function (param) {
    return Promise.resolve(22);
  };
  var obj = p.catch(arg1);
  var arg1$1 = function (value) {
    return Promise.resolve(assert_bool(value === 42));
  };
  var obj$1 = obj.then(arg1$1);
  return obj$1.catch(fail);
}

function orRejectedTest(param) {
  var p = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var arg1 = function (param) {
    return Promise.resolve(22);
  };
  var obj = p.catch(arg1);
  var arg1$1 = function (value) {
    return Promise.resolve(assert_bool(value === 22));
  };
  var obj$1 = obj.then(arg1$1);
  return obj$1.catch(fail);
}

function orElseResolvedTest(param) {
  var p = Promise.resolve(42);
  var arg1 = function (param) {
    return Promise.resolve(22);
  };
  var obj = p.catch(arg1);
  var arg1$1 = function (value) {
    return Promise.resolve(assert_bool(value === 42));
  };
  var obj$1 = obj.then(arg1$1);
  return obj$1.catch(fail);
}

function orElseRejectedResolveTest(param) {
  var p = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var arg1 = function (param) {
    return Promise.resolve(22);
  };
  var obj = p.catch(arg1);
  var arg1$1 = function (value) {
    return Promise.resolve(assert_bool(value === 22));
  };
  var obj$1 = obj.then(arg1$1);
  return obj$1.catch(fail);
}

function orElseRejectedRejectTest(param) {
  var p = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var arg1 = function (param) {
    return Promise.reject({
                RE_EXN_ID: "Stack_overflow"
              });
  };
  var obj = p.catch(arg1);
  var obj$1 = obj.then(fail);
  var arg1$1 = function (error) {
    var match = Caml_exceptions.is_extension(error) && error.RE_EXN_ID === "Stack_overflow" ? 0 : undefined;
    if (match !== undefined) {
      return h;
    }
    throw {
          RE_EXN_ID: "Assert_failure",
          _1: [
            "js_promise_basic_test.ml",
            77,
            18
          ],
          Error: new Error()
        };
  };
  return obj$1.catch(arg1$1);
}

function resolveTest(param) {
  var p1 = Promise.resolve(10);
  var arg1 = function (x) {
    return Promise.resolve(assert_bool(x === 10));
  };
  return p1.then(arg1);
}

function rejectTest(param) {
  var p = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var arg1 = assertIsNotFound;
  return p.catch(arg1);
}

function thenCatchChainResolvedTest(param) {
  var p = Promise.resolve(20);
  var arg1 = function (value) {
    return Promise.resolve(assert_bool(value === 20));
  };
  var obj = p.then(arg1);
  return obj.catch(fail);
}

function thenCatchChainRejectedTest(param) {
  var p = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var obj = p.then(fail);
  var arg1 = assertIsNotFound;
  return obj.catch(arg1);
}

function allResolvedTest(param) {
  var p1 = Promise.resolve(1);
  var p2 = Promise.resolve(2);
  var p3 = Promise.resolve(3);
  var promises = [
    p1,
    p2,
    p3
  ];
  var obj = Promise.all(promises);
  var arg1 = function (resolved) {
    assert_bool(Caml_array.get(resolved, 0) === 1);
    assert_bool(Caml_array.get(resolved, 1) === 2);
    assert_bool(Caml_array.get(resolved, 2) === 3);
    return h;
  };
  return obj.then(arg1);
}

function allRejectTest(param) {
  var p1 = Promise.resolve(1);
  var p2 = Promise.resolve(3);
  var p3 = Promise.reject({
        RE_EXN_ID: "Not_found"
      });
  var promises = [
    p1,
    p2,
    p3
  ];
  var obj = Promise.all(promises);
  var obj$1 = obj.then(fail);
  var arg1 = function (error) {
    assert_bool(error === ({
            RE_EXN_ID: "Not_found"
          }));
    return h;
  };
  return obj$1.catch(arg1);
}

function raceTest(param) {
  var p1 = Promise.resolve("first");
  var p2 = Promise.resolve("second");
  var p3 = Promise.resolve("third");
  var promises = [
    p1,
    p2,
    p3
  ];
  var obj = Promise.race(promises);
  var arg1 = function (resolved) {
    return h;
  };
  var obj$1 = obj.then(arg1);
  return obj$1.catch(fail);
}

function createPromiseRejectTest(param) {
  var obj = new Promise((function (resolve, reject) {
          return reject({
                      RE_EXN_ID: "Not_found"
                    });
        }));
  var arg1 = function (error) {
    assert_bool(error === ({
            RE_EXN_ID: "Not_found"
          }));
    return h;
  };
  return obj.catch(arg1);
}

function createPromiseFulfillTest(param) {
  var obj = new Promise((function (resolve, param) {
          return resolve("success");
        }));
  var arg1 = function (resolved) {
    assert_bool(resolved === "success");
    return h;
  };
  var obj$1 = obj.then(arg1);
  return obj$1.catch(fail);
}

thenTest(undefined);

andThenTest(undefined);

catchTest(undefined);

orResolvedTest(undefined);

orRejectedTest(undefined);

orElseResolvedTest(undefined);

orElseRejectedResolveTest(undefined);

orElseRejectedRejectTest(undefined);

thenCatchChainResolvedTest(undefined);

thenCatchChainRejectedTest(undefined);

allResolvedTest(undefined);

allRejectTest(undefined);

raceTest(undefined);

createPromiseRejectTest(undefined);

createPromiseFulfillTest(undefined);

var obj = Promise.all([
      Promise.resolve(2),
      Promise.resolve(3)
    ]);

function arg1(param) {
  eq("File \"js_promise_basic_test.ml\", line 169, characters 7-14", [
        param[0],
        param[1]
      ], [
        2,
        3
      ]);
  return Promise.resolve(undefined);
}

obj.then(arg1);

console.log(List.length(suites.contents));

console.log("hey");

Mt.from_pair_suites("Js_promise_basic_test", suites.contents);

var twop = Promise.resolve(2);

function re(prim) {
  return Promise.resolve(prim);
}

Mt.from_promise_suites("Js_promise_basic_test", {
      hd: [
        "File \"js_promise_basic_test.ml\", line 187, characters 4-11",
        twop.then(function (x) {
              return Promise.resolve({
                          TAG: /* Eq */0,
                          _0: x,
                          _1: 2
                        });
            })
      ],
      tl: {
        hd: [
          "File \"js_promise_basic_test.ml\", line 190, characters 4-11",
          twop.then(function (x) {
                return Promise.resolve({
                            TAG: /* Neq */1,
                            _0: x,
                            _1: 3
                          });
              })
        ],
        tl: /* [] */0
      }
    });

var then_ = Js_promise.then_;

exports.suites = suites;
exports.test_id = test_id;
exports.eq = eq;
exports.assert_bool = assert_bool;
exports.fail = fail;
exports.thenTest = thenTest;
exports.andThenTest = andThenTest;
exports.h = h;
exports.assertIsNotFound = assertIsNotFound;
exports.catchTest = catchTest;
exports.orResolvedTest = orResolvedTest;
exports.orRejectedTest = orRejectedTest;
exports.orElseResolvedTest = orElseResolvedTest;
exports.orElseRejectedResolveTest = orElseRejectedResolveTest;
exports.orElseRejectedRejectTest = orElseRejectedRejectTest;
exports.resolveTest = resolveTest;
exports.rejectTest = rejectTest;
exports.thenCatchChainResolvedTest = thenCatchChainResolvedTest;
exports.thenCatchChainRejectedTest = thenCatchChainRejectedTest;
exports.allResolvedTest = allResolvedTest;
exports.allRejectTest = allRejectTest;
exports.raceTest = raceTest;
exports.createPromiseRejectTest = createPromiseRejectTest;
exports.createPromiseFulfillTest = createPromiseFulfillTest;
exports.twop = twop;
exports.then_ = then_;
exports.re = re;
/* h Not a pure module */
